precompile/warp_messenger.go

```diff
@@ -0,0 +1,444 @@
+	SubmitMessageEventID = "da2b1cd3e6664863b4ad90f53a4e14fca9fc00f3f0e01e5c7b236a4355b6591a" // Keccack256("SubmitMessage(bytes32,uint256)")
+)
+
+// Reference imports to suppress errors from unused imports. This code and any unnecessary imports can be removed.
+var (
+	_ = errors.New
+	_ = big.NewInt
+)
+
+// Singleton StatefulPrecompiledContract and signatures.
+var (
+	_ StatefulPrecompileConfig = &WarpMessengerConfig{}
+
+	WarpMessengerABI        abi.ABI                     // will be initialized by init function
+	WarpMessengerPrecompile StatefulPrecompiledContract // will be initialized by init function
+
+	// Default stake threshold for aggregate signature verification. (67%)
+	WarpQuorumNumerator   = uint64(67)
+	WarpQuorumDenominator = uint64(100)
+
+	signedMessages []*warp.Message
+
+	ErrMissingStorageSlots       = errors.New("missing access list storage slots from precompile during execution")
+	ErrInvalidMessageIndex       = errors.New("invalid message index")
+	ErrInvalidSignature          = errors.New("invalid aggregate signature")
+	ErrMissingProposerVMBlockCtx = errors.New("missing proposer VM block context")
+	ErrWrongChainID              = errors.New("wrong chain id")
+	ErrInvalidQuorumDenominator  = errors.New("quorum denominator can not be zero")
+	ErrGreaterQuorumNumerator    = errors.New("quorum numerator can not be greater than quorum denominator")
+)
+
+// WarpMessengerConfig implements the StatefulPrecompileConfig
+// interface while adding in the WarpMessenger specific precompile address.
+type WarpMessengerConfig struct {
+	UpgradeableConfig
+	InitialWarpConfig *InitialWarpConfig `json:"initialWarpConfig,omitempty"`
+}
+
+// WarpMessage is an auto generated low-level Go binding around an user-defined struct.
+type WarpMessage struct {
+	OriginChainID       [32]byte
+	OriginSenderAddress [32]byte
+	DestinationChainID  [32]byte
+	DestinationAddress  [32]byte
+	Payload             []byte
+}
+
+type GetVerifiedWarpMessageOutput struct {
+	Message WarpMessage
+	Success bool
+}
+
+type SendWarpMessageInput struct {
+	DestinationChainID [32]byte
+	DestinationAddress [32]byte
+	Payload            []byte
+}
+
+type InitialWarpConfig struct {
+	QuorumNumerator   uint64 `json:"quorumNumerator,omitempty"`
+	QuorumDenominator uint64 `json:"quorumDenominator,omitempty"`
+}
```

Comment:
The prefix `Initial` has been used when we want to take an initial action like mint funds or add specific addresses to the allow list. Can we remove this and include these parameters directly in `WarpConfig` ?